### PR TITLE
Fix bailout broken in lazy components due to default props resolving

### LINKED ISSUE #17151: Bailing out doesn't work properly in lazy components with default props
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Bailing out doesn't work properly in lazy components with default props. It seems we're incorrectly [comparing unresolved props (oldProps) with resolved props (newProps)](https://github.com/facebook/react/blob/62b04cfa753076d5ffb1d74b855f8f8db36f5186/packages/react-reconciler/src/ReactFiberClassComponent.js#L1100).

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**
https://codesandbox.io/s/stoic-curran-3otbb

**What is the expected behavior?**
In the example above, `componentDidUpdate` shouldn't have been called when the button is clicked.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

### PR DESCRIPTION
**Pull request by jddxf**
Fix #17151 

We should never compare unresolved props with resolved props. Since comparing resolved props by reference doesn't make sense, we use unresolved props in that case. Otherwise, resolved props are used.

### CODE REVIEW

**Comment by bl00mber** [comment]
@jddxf you probably may want to change the way you pull updated changes form react original repo to your fork because when you pull them you leave a lot of references in all the other pull requests!!!

Screenshot from 2020-04-05 1

Screenshot from 2020-04-05 2

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 1]
```diff
@@ -1056,20 +1061,21 @@ function updateClassInstance(
...
     if (typeof instance.getSnapshotBeforeUpdate === 'function') {
       if (
-        oldProps !== current.memoizedProps ||
+        unresolvedOldProps !== current.memoizedProps ||
         oldState !== current.memoizedState
       ) {
         workInProgress.effectTag |= Snapshot;
       }
     }
+    instance.props = workInProgress.memoizedProps = newProps;
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberClassComponent.js:1078]: Why is this needed?

**Comment by jddxf** [comment]
@bl00mber Sorry for that. I did force pushed the right commits but that didn't work. I had to delete the fork and it seems ok now.

File: packages/react-reconciler/src/ReactFiberClassComponent.js (hunk 1 above)
Comment (jddxf) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1078]: Otherwise, React would warn.
https://github.com/facebook/react/blob/1fd45437d781b1b905be2308447aa7793bbcb571/packages/react-reconciler/src/ReactFiberBeginWork.js#L859-L868

This also mirrors the logic at the end of the same function.
https://github.com/facebook/react/blob/1fd45437d781b1b905be2308447aa7793bbcb571/packages/react-reconciler/src/ReactFiberClassComponent.js#L1139-L1151

This was not needed before because it was already guaranteed by the condition in the branch.
https://github.com/facebook/react/blob/1fd45437d781b1b905be2308447aa7793bbcb571/packages/react-reconciler/src/ReactFiberClassComponent.js#L1050

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 2]
```diff
@@ -997,11 +997,13 @@ function updateClassInstance(
 
   cloneUpdateQueue(current, workInProgress);
 
-  const oldProps = workInProgress.memoizedProps;
-  instance.props =
+  const unresolvedOldProps = workInProgress.memoizedProps;
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberClassComponent.js:1000]: This says they're unresolved. But later we do this:

```js
instance.props = workInProgress.memoizedProps = newProps;
```

So next time we *would* have resolved props in `memoizedProps`? I don't quite follow the logic.

Comment (jddxf) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1000]: Since we reset `workInProgress.memoizedProps` to `workInProgress.pendingProps` right after begin work, `workInProgress.memoizedProps` is always unresolved when we begin work.
https://github.com/facebook/react/blob/e69ca310ea56b6f2f510dd463902405285953432/packages/react-reconciler/src/ReactFiberWorkLoop.js#L1501

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1000]: Then can we assign `instance.props` only? Is there a reason for `workInPorgress.memoizedProps` assignment? Sorry I'm being dense, it's been a while since I looked a tthis.

Comment (jddxf) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1000]: Never mind. But unfortunately we have to keep their values in sync. [Here](https://github.com/facebook/react/pull/18491#discussion_r403638636) is the reason.

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1000]: I still don't think it makes sense to assign either of them.

In the sCU branch, `instance.props` reassignment makes sense because the passed props actually *did* change. We want the most recent values inside of them to be accessible to the lifecycle methods, even if sCU "lies" about them not having changed.

In this branch, we've verified props are `===`. So the only reason they "changed" is because we resolved default props. That's not a compelling reason to give our components a different props object.

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1000]: If you change the warning condition to `if (shouldUpdate && inst.props !== nextProps) {` then the warning won't fire. It will also make sense because *the render wasn't called* — so the warning message would have been misleading anyway.

Comment (jddxf) [reply packages/react-reconciler/src/ReactFiberClassComponent.js:1000]:
```
instance.props = workInProgress.memoizedProps = newProps;
```
This only gives our components a different props object when we've actually resolved default props. But in that case, components are already given a new props object.
https://github.com/facebook/react/blob/2dddd1e00cdadad1783aa55ac65a28634590e9c1/packages/react-reconciler/src/ReactFiberClassComponent.js#L1001-L1004
I think it's ok to remove this assignment if keeping `workInProgress.memoizedProps` and `instance.props` stay the same until beginWork finishes is not a **requirement**. And that seems true now. So let's change the warning condition instead.

**Comment by jddxf** [comment]
I cannot update this PR now due to refork. Closed in favour of #18539.
